src/rewritable_units/text_encoder.rs

```diff
@@ -0,0 +1,460 @@
+    pub fn new() -> Self {
+        Self {
+            char_bytes: [0; 4],
+            char_len: 0,
+        }
+    }
+
+    /// Returns a valid UTF-8 fragment, and not-yet-checked remainder of the bytes.
+    ///
+    /// Call `discard_incomplete()` after the last write to flush any partially-written chars.
+    pub fn utf8_bytes_to_slice<'buf, 'src: 'buf>(
+        &'buf mut self,
+        mut content: &'src [u8],
+    ) -> Result<(&'buf str, &'src [u8]), Utf8Error> {
+        // There may be incomplete char buffered from previous write, that must be continued now
+        if self.char_len > 0 {
+            let mut must_emit_now = false;
+            while let Some((&next_byte, rest)) = content.split_first() {
+                if is_continuation_byte(next_byte) {
+                    if let Some(buf) = self.char_bytes.get_mut(self.char_len as usize) {
+                        *buf = next_byte;
+                        self.char_len += 1;
+                        content = rest;
+                        continue;
+                    }
+                    // overlong sequences fall here, and will be checked when the char_bytes is flushed
+                }
+                must_emit_now = true;
+                break;
+            }
+
+            if self.char_len >= utf8_width(self.char_bytes[0]) {
+                must_emit_now = true;
+            }
+
+            if must_emit_now {
+                let char_buf = self
+                    .char_bytes
+                    .get(..self.char_len as usize)
+                    .ok_or(Utf8Error)?;
+                self.char_len = 0;
+                let ch = std::str::from_utf8(char_buf).map_err(|_| Utf8Error)?;
+                Ok((ch, content))
+            } else {
+                // a partial write has ended without fully completing a char (it's possible to write 1 byte at a time)
+                debug_assert!(content.is_empty());
+                Ok(("", b""))
+            }
+        } else {
+            match std::str::from_utf8(content) {
+                Ok(src) => Ok((src, b"")),
+                // error_len means invalid bytes somewhere, not just incomplete 1-3 bytes at the end
+                Err(err) if err.error_len().is_some() => Err(Utf8Error),
+                Err(err) => {
+                    let (valid, invalid) = content
+                        .split_at_checked(err.valid_up_to())
+                        .ok_or(Utf8Error)?;
+                    // save the incomplete bytes from the end for the next write
+                    self.char_bytes
+                        .get_mut(..invalid.len())
+                        .ok_or(Utf8Error)?
+                        .copy_from_slice(invalid);
+                    self.char_len = invalid.len() as _;
+                    // valid_up_to promises it is valid
+                    debug_assert!(std::str::from_utf8(valid).is_ok());
+                    let valid = unsafe { std::str::from_utf8_unchecked(valid) };
```

Comment:
Can you call `std::str::from_utf8()` instead and `.except("...")` on it (and remove the `debug_assert!()` above). I prefer to keep `unsafe` to an absolute minimum to have the least possible surface for UB.

Reply:
Replaced with safe non-panicking code